src/lib/field.ts

```diff
@@ -0,0 +1,1194 @@
+    if (Field.#isField(x)) {
+      this.value = x.value;
+      return;
+    }
+    // fieldVar
+    if (Array.isArray(x)) {
+      this.value = x;
+      return;
+    }
+    // TODO this should handle common values efficiently by reading from a lookup table
+    this.value = FieldVar.constant(Fp(x));
+  }
+
+  // helpers
+  static #isField(x: bigint | number | string | Field | FieldVar): x is Field {
+    return x instanceof Field || (x as any) instanceof SnarkyFieldConstructor;
+  }
+  static #toConst(x: bigint | number | string | ConstantField): FieldConst {
+    if (Field.#isField(x)) return x.value[1];
+    return FieldConst.fromBigint(Fp(x));
+  }
+  static #toVar(x: bigint | number | string | Field): FieldVar {
+    if (Field.#isField(x)) return x.value;
+    return FieldVar.constant(Fp(x));
+  }
+  static from(x: bigint | number | string | Field): Field {
+    if (Field.#isField(x)) return x;
+    return new Field(x);
+  }
+
+  /**
+   * Check whether this {@link Field} element is a hard-coded constant in the constraint system.
+   * If a {@link Field} is constructed outside a zkApp method, it is a constant.
+   *
+   * @example
+   * ```ts
+   * console.log(Field(42).isConstant()); // true
+   * ```
+   *
+   * @example
+   * ```ts
+   * \@method myMethod(x: Field) {
+   *    console.log(x.isConstant()); // false
+   * }
+   * ```
+   *
+   * @return A `boolean` showing if this {@link Field} is a constant or not.
+   */
+  isConstant(): this is { value: ConstantFieldVar } {
+    return this.value[0] === FieldType.Constant;
+  }
+
+  /**
+   * Create a {@link Field} element equivalent to this {@link Field} element's value,
+   * but is a constant.
+   * See {@link Field.isConstant} for more information about what is a constant {@link Field}.
+   *
+   * @example
+   * ```ts
+   * const someField = Field(42);
+   * someField.toConstant().assertEquals(someField); // Always true
+   * ```
+   *
+   * @return A constant {@link Field} element equivalent to this {@link Field} element.
+   */
+  toConstant(): ConstantField {
+    if (this.isConstant()) return this;
+    // TODO: fix OCaml error message, `Can't evaluate prover code outside an as_prover block`
+    let value = Snarky.field.readVar(this.value);
+    return new Field(FieldVar.constant(value)) as ConstantField;
+  }
+
+  /**
+   * Serialize the {@link Field} to a bigint, e.g. for printing. Trying to print a {@link Field} without this function will directly stringify the Field object, resulting in an unreadable output.
+   *
+   * **Warning**: This operation does _not_ affect the circuit and can't be used to prove anything about the bigint representation of the {@link Field}. Use the operation only during debugging.
+   *
+   * @example
+   * ```ts
+   * const someField = Field(42);
+   * console.log(someField.toBigInt());
+   * ```
+   *
+   * @return A bigint equivalent to the bigint representation of the Field.
+   */
+  toBigInt() {
```

Comment:
can we change this `debugToBigInt()` or `toBigIntDebug()` something 
 like that to make it clear in the name that it's only for debugging?

We've seen so many people make this mistake, it's probably worth it.

I imagine tactically, you'd need to deprecate these since they're in-use already and we can remove them fully later.

Reply:
> Instead, everything that's not a variable should just be a plain JS type, like bigint or number, and it should be possible to use that in place of constant Field elements everywhere

I like this ^^, especially because you can monkey-patch things in JavaScript (though of course we should be careful)